sql: fix UPSERT with columns after partial index PUT and DEL columns

mgartner opened the pull request:
#### sql: fix UPSERT with columns after partial index PUT and DEL columns

This commit fixes a bug in UPSERT execution that was caused by the
incorrect assumption that there are never columns following the partial
index PUT and DEL columns in the mutation's input.

I was unable to reproduce this bug for INSERTs or DELETEs. However, I
updated the execution code for both out of caution. The logic for both
no longer makes the same assumption that partial index PUT and DEL
columns will be the last columns in the input.

Fixes #61414

Release justification: This is a low-risk fix for a bug causing UPSERT
statements to err when executed on tables with partial indexes.

Release note (bug fix): A bug which caused UPSERT and INSERT..ON
CONFLICT..DO UPDATE statements to fail on tables with both partial
indexes and foreign key references has been fixed. This bug has been
present since version 20.2.0.

#### ccl: add tests for implicitly partitioned partial unqiue indexes

Release justification: This is a test-only change.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/61416)
<!-- Reviewable:end -->

rytaft reviewed (commented):
Good find!

Is there nothing else that triggers this bug? It would be great if you could find another regression test, since I'm not convinced the test cases in `pkg/ccl/logictestccl` are fully exercising it (we don't plan uniqueness checks for DELETE, for example...) 

[nit] the commit/PR title mentions nothing about the bug fix, which seems more important than the test cases....

> Reviewed 6 of 6 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/61416#-:-MUtsCbPE2lhA93MgB1L:b-bu0a3a)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @RaduBerinde)



<!-- Sent from Reviewable.io -->

RaduBerinde reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/61416#-:-MUtw-kABxddLEWXhgeY:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained



<!-- Sent from Reviewable.io -->

mgartner reviewed (commented):
I updated the commit message and PR title. I also added a simplified regression test that doesn't require a RBR table.

I was unable to reproduce for DELETE or INSERT. This is likely because this problem doesn't present when columns are fully pruned, which currently is done correctly for FK cascades but not uniqueness checks (until #61376 is merged). Therefore, we can't reproduce for DELETEs because they would only have FK cascades, never uniqueness checks. And we can't reproduce for INSERTs because they always include all rows in the mutation input, so there is nothing to prune. But I think it's smart to make these changes to `insert.go` and `delete.go` to be extra cautious.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/61416#-:-MUuMbm21kP8CqEDM9Kz:blfeuls)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @rytaft)



<!-- Sent from Reviewable.io -->

rytaft reviewed (approved):
Makes sense - thanks!

<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

> Reviewed 2 of 2 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/61416#-:-MUuOtNUCEAUPcsfv3uW:b-s5xfgf)* status: :shipit: complete! 1 of 0 LGTMs obtained (and 1 stale)



<!-- Sent from Reviewable.io -->

mgartner committed:
sql: fix UPSERT with columns after partial index PUT and DEL columns

This commit fixes a bug in UPSERT execution that was caused by the
incorrect assumption that there are never columns following the partial
index PUT and DEL columns in the mutation's input.

I was unable to reproduce this bug for INSERTs or DELETEs. However, I
updated the execution code for both out of caution. The logic for both
no longer makes the same assumption that partial index PUT and DEL
columns will be the last columns in the input.

Fixes #61414

Release justification: This is a low-risk fix for a bug causing UPSERT
statements to err when executed on tables with partial indexes.

Release note (bug fix): A bug which caused UPSERT and INSERT..ON
CONFLICT..DO UPDATE statements to fail on tables with both partial
indexes and foreign key references has been fixed. This bug has been
present since version 20.2.0.

mgartner committed:
ccl: add tests for implicitly partitioned partial unqiue indexes

Release justification: This is a test-only change.

Release note: None

mgartner commented:
@rytaft or @RaduBerinde PTAL. I split the bug fix and multi-region changes into two commits to make the bug fix more easily backport-able. @karlseguin ran into this bug in 20.2.5, proving it's not only producible with multi-region features, so I added a simplified version of his reproduction to the regression tests.

rytaft reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

> Reviewed 3 of 3 files at r3, 2 of 2 files at r4.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/61416#-:-MUyNDhEBF16rEpgIFNA:bg7igkv)* status: :shipit: complete! 1 of 0 LGTMs obtained (and 1 stale)



<!-- Sent from Reviewable.io -->

RaduBerinde reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/61416#-:-MUySY-46MDZRBqyhzsk:bnfp4nl)* status: :shipit: complete! 2 of 0 LGTMs obtained



<!-- Sent from Reviewable.io -->

mgartner commented:
TFTRs!

bors r+
